### PR TITLE
Turn order

### PR DESCRIPTION
**Pull request by DanUgore**
This is just to show exactly what I added for turn order. Not to be merged.

### CODE REVIEW

**Comment by DanUgore** [comment]
@Zarel Here it is, seems fully functional now for actions. Haven't looked at sorting residuals with this yet.

File: battle-engine.js [hunk 1]
```diff
@@ -3173,6 +3177,7 @@ Battle = (function () {
 			if (!decision.pokemon && !decision.speed) decision.speed = 1;
 			if (!decision.speed && decision.choice === 'switch' && decision.target) decision.speed = decision.target.speed;
 			if (!decision.speed) decision.speed = decision.pokemon.speed;
+			if (decision.choice === 'move' || decision.choice === 'switch') decision.speed = decision.speed & 8191; // Silly GameFreak shenanigans.
```
Comment (DanUgore) [inline battle-engine.js:3181]: Do you think I should hardcode Trick Room here btw?

**Comment by Zarel** [comment]
I still don't know what's wrong with just using a stable sort. :|

**Comment by Zarel** [comment]
Why do we even sort mid-turn, anyway? In theory, it should never happen.

**Comment by DanUgore** [comment]
Sorry I still don't understand how to sort stably here. When we sort we need to resolve speed-ties, which we cant do if we sort before all pokemon are present.

**Comment by Zarel** [comment]
So, why can't we sort after all pokemon are present?

**Comment by DanUgore** [comment]
This does. In the 'start' event, `switchIn()` would sort after every call and it was called once for every Pokemon as it came in on Turn 0. That's no longer an issue,

As for sorting mid-turn, there is the move Quash to deal with. That's all I can think of.

**Comment by Zarel** [comment]
Quash can manually change turn order, it doesn't need to trigger a full sort.

**Comment by DanUgore** [comment]
Yea I just saw that

**Comment by Zarel** [comment]
Basically, sort should happen at the beginning of every turn, and right after all Pokémon are switched in during turn 0. `BattlePokemon.prototype.tieOrder` should not exist. Anything else?

**Comment by DanUgore** [comment]
Wait, why should that not exist?

**Comment by DanUgore** [comment]
Well I guess it's actually possible/better to keep track of that in the battle object as opposed to each pokemon/

**Comment by Zarel** [comment]
No, I mean, you should never have to keep track of it, since we should only ever sort once, at which point ties should be resolved instantly, right?

**Comment by DanUgore** [comment]
Oh I see, well there's that and we need sort in the same order for end-of-turn residuals.

**Comment by Zarel** [comment]
Oooooooh.

Yeah, keep `.tieOrder`.

**Comment by Zarel** [comment]
Wait, how do switches work re: residuals?

**Comment by DanUgore** [comment]
That is one of the things on my "to investigate" list.

**Comment by Zarel** [comment]
You might want to quash this. At least the "fix" commits.

**Comment by DanUgore** [comment]
K
- V4

**Comment by DanUgore** [comment]
Well apparently there is a lot of rerolling of speed ties.
1) Before the turn plays out
2) After Mega Evolution
3) Before residuals

So I wonder if we even need to keep track of tieOrder if we don't resort without rerolling after it's dealt with.

**Comment by DanUgore** [comment]
Ok so it looks like we can get rid of tie order. Everywhere it's resorted it rerolls so we'll see how this turns out.

**Comment by Zarel** [comment]
Okay, good, that makes things simpler.

**Comment by Zarel** [comment]
So... is there anything that still needs to be done here?

**Comment by DanUgore** [comment]
Thanks for reminding me. Mega evolution needs to be it's own queue thing. Speed is refilled post-mega evolution. Winning the mega speed tie doesn't guarantee you the same when you move.

**Comment by DanUgore** [comment]
Will probably look into this again. The fact Im actually approached a few times about double speed ties to this day is actually kind of dumb. Lost the local repo so Ill have to do some work over.

**Pull request closed without merging**